Issue/11750 dashboard big screen support

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13079
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Based on the changes I worked on here: pe5sF9-3si-p2 and resulting in this [PR](https://github.com/woocommerce/woocommerce-android/pull/13079) work, I've refactored the prior implementation to avoid using `LazyStaggeredGrid` to support big screen in the dashboard tab. Instead, this relies on a custom implementation that will manually add 2 or 3 columns to the dashboard based on the screen's `currentWindowAdaptiveInfo().windowSizeClass`. 

Changes added: 

- Removed usage of xml `NestedScrollView`. This was needed in order to add Composable colums with vertical scroll that work well/coordinate with the collapsing toolbar.
- Refactored pull to refresh implementation to use Composable approach instead of relying in our custom view. 
- Adjusted columns and manually created grid to coordinate with the collapsing toolbar. 


There's one thing I'm not fully convinced. To determine the number of columns, I'm using [windowSizeClass.windowWidthSizeClass](https://developer.android.com/develop/ui/compose/layouts/adaptive/use-window-size-classes).

![Screenshot 2025-02-07 at 09 13 45](https://github.com/user-attachments/assets/2ae46165-4463-4334-89a3-0ececef8bdbc)

This results in bigger phones showing 3 columns in landscape (see screen recordings below). I don't think it is a major issue, but I'd say two columns would look better. Let me know what you think, although I suspect it won't be straightforward to force two columns for mobile phones on landscape mode. 

### Testing information
Just run the app on the phone screen and tablet screen and check that the dashboard works as expected. 
- Rotate the device and test the number of columns adjusts accordingly
- Check pull to refresh works as expected
- Check scrolling up and down works well with the collapsing toolbar
- Add and remove cards at will and check that they are displayed in the right order. 

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
The above ☝🏼 

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

**Tablet Experience**

https://github.com/user-attachments/assets/c12b4bb5-2444-48ad-a56b-221eed90071e


**Large phone Eexperience (Pixel 8 pro)**

https://github.com/user-attachments/assets/13afdfbf-8422-468e-b6a5-238696d658a6


**Smaller phone experience**


https://github.com/user-attachments/assets/6af0745c-0dda-4b49-8170-7a8ebc1d6ee2


- [X] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Adds grid layout for bigger screens

JorgeMucientes committed:
Extract dashboard container out of the xml NestedScroll and SwipeToRefresh

JorgeMucientes committed:
Add a composable implementation of swipe to refresh

JorgeMucientes committed:
Clean xml layout and ensure we keep the same logic in the DashboardFragment

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support

JorgeMucientes committed:
Refactor: Adapt Dashboard card columns to Window Size Classes

- Replace custom `WindowSizeClass` with `WindowWidthSizeClass` from `androidx.window.core.layout`.
- Use `currentWindowAdaptiveInfo` instead custom extension that ignores screen width to calculate SizeClass
- Update the number of columns in the Dashboard based on the Window Size Class.
- Use a `Column` with vertical scrolling instead of `LazyColumn` when `numberOfColumns` is 1.
- When multiple columns in a row instead of LazyVerticalStaggeredGrid that doesn't work well with nested scrolling interactions
- Adds dependency to adaptiveAndroid

JorgeMucientes committed:
Fix unit tests after adding composable pull to refresh implementation

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes committed:
Remove animation when cards are displayed in multiple columns

JorgeMucientes committed:
Update release notes

JorgeMucientes committed:
Adjust dashboard card paddings

JorgeMucientes commented on WooCommerce/build.gradle:
```diff
@@ -411,6 +411,7 @@ dependencies {
     implementation(libs.androidx.compose.material.icons.extended)
     implementation(libs.androidx.compose.ui.text.google.fonts)
     implementation(libs.androidx.navigation.compose)
+    implementation(libs.androidx.adaptive.android)
```
Added this dependency to rely on official library to categorize screen size. We currently have a custom implmentation of [WindowSizeClass](https://github.com/woocommerce/woocommerce-android/blob/c4dd29b8c60f199de5407bccb8c503cde2eb2554/WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/ContextExt.kt#L56) but I think we should be using the official one. Also the custom implementation of `WindowSizeClass` currently only exposes `windowHeightSizeClass` but not `windowWidthSizeClass`. So instead of manually adding `windowWidthSizeClass`, I figured we could use[ the official size class library for it.](https://developer.android.com/develop/ui/compose/layouts/adaptive/use-window-size-classes)

JorgeMucientes commented:
Setting the milestone to `21.8`. I noticed the code freeze hasn't happened yet. I don't want to rush this into `trunk` and rather leave more time for testing, as there are multiple changes in one of the main screens of the app.

pachlava reviewed (changes requested):
@JorgeMucientes 👋 

I noticed that this fails:
```
> Task :WooCommerce:compileWasabiDebugAndroidTestKotlin FAILED

e: file:///Users/.../woocommerce-android/WooCommerce/src/androidTest/kotlin/com/woocommerce/android/e2e/screens/mystore/DashboardScreen.kt:17:37 Unresolved reference 'my_store_refresh_layout'.
```

Making tests execution impossible. This is caused by the removal of `my_store_refresh_layout` from `fragment_dashboard.xml` and probably can be fixed by replacing `R.id.my_store_refresh_layout` in `DashboardScreen.kt:17:37` by another component (ideally, unique), the presence of which can inform that we're on the Dashboard Screen.

pachlava committed:
Fix the unexisting UI component reference.

pachlava reviewed (approved):
@JorgeMucientes I took a liberty of updating the missing reference in UI tests with 581d2e9cb9ec6e32558d90775a79b04d3be19fb9, so tests-wise it's good to go now

JorgeMucientes commented:
> @JorgeMucientes I took a liberty of updating the missing reference in UI tests with https://github.com/woocommerce/woocommerce-android/commit/581d2e9cb9ec6e32558d90775a79b04d3be19fb9, so tests-wise it's good to go now

Awesome @pachlava. Thanks so much for proactively taking a look into this 🥇

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support

JorgeMucientes committed:
Update release notes

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardContainer.kt:
```diff
@@ -45,70 +60,98 @@ import com.woocommerce.android.ui.dashboard.stock.DashboardProductStockCard
+                    .fillMaxSize()
+                    .background(MaterialTheme.colors.surface)
+                    .padding(16.dp),
+                numberOfColumns = when (windowSizeClass.windowWidthSizeClass) {
+                    WindowWidthSizeClass.COMPACT -> 1
+                    WindowWidthSizeClass.MEDIUM -> 2
+                    WindowWidthSizeClass.EXPANDED -> 3
+                    else -> 1
```
Should we coerce the numberOfColumns to the number of available cards (`coerceIn(maximumValue = state.widgets.size)`? For example if I have only one card, and I use the app in landscape, it will appear like this:

![Screenshot_20250210_185256](https://github.com/user-attachments/assets/2d57b59e-6d9f-49be-bb85-da9eff7624ba)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardContainer.kt:
```diff
@@ -45,70 +60,98 @@ import com.woocommerce.android.ui.dashboard.stock.DashboardProductStockCard
+                    .fillMaxSize()
+                    .background(MaterialTheme.colors.surface)
+                    .padding(16.dp),
+                numberOfColumns = when (windowSizeClass.windowWidthSizeClass) {
```
Should we coerce the numberOfColumns to the number of available cards (`coerceIn(maximumValue = state.widgets.size)`? For example if I have only one card, and I use the app in landscape, it will appear like this:

![Screenshot_20250210_185256](https://github.com/user-attachments/assets/2d57b59e-6d9f-49be-bb85-da9eff7624ba)

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:
```diff
@@ -362,7 +349,11 @@ class DashboardFragment :
-    override fun shouldExpandToolbar() = binding.statsScrollView.scrollY == 0
+    override fun shouldExpandToolbar() = true
+
+    override fun scrollToTop() {
+        return
+    }
```
Should we add a TODO about bringing back this functionality? This was called when re-selecting the tab, but it's broken now, as `MainActivity#getActiveTopLevelFragment` seems broken in `trunk`, I will try to investigate why later.

JorgeMucientes replied:
Hmm, good point. I removed the nestedScrollView on XML I'm not sure how feasible would it be to fix this. Going to investigate it, if its a quick change, I'll apply it in this PR. If not, I'll create a new GH issue.

hichamboushaba replied:
One way would be to host the `ScrollState` in the ViewModel (or the Fragment if we can easily restore its state), and then pass it to the Composable code from there, and use its function (`animateScrollTo` or `scrollTo`) for triggering the scroll to top.

JorgeMucientes replied:
So I look a bit into this, and fixing it is more complicated than I though. 
I did some quick and dirty changes to try to fix `shouldExpandToolbar` and `scrollToTop` from `DashboardFragment` and noticed the following issues: 
- `scrollToTop()` is never called when re-selecting the tab. This happens to all the other tabs as well and it is because  `getActiveTopLevelFragment()` returns `null` always, when reselecting any tab. I don't know since when this is broken and haven't had the time to dig deper
Regarding `shouldExpandToolbar,` I think there's a race condition where `shouldExpandToolbar()` is called when the toolbar is already expanded (too late). This causes the toolbar to always be in expanded mode when navigating back to `My Store` tabs from other tabs, regardless of the `scrollState` value. 

Created a [GH issue](https://github.com/woocommerce/woocommerce-android/issues/13524) but I won't have time to look into this shortly.

hichamboushaba reviewed (commented):
Really nice work @JorgeMucientes 👏, code looks good, but:

> There's one thing I'm not fully convinced. To determine the number of columns, I'm using [windowSizeClass.windowWidthSizeClass](https://developer.android.com/develop/ui/compose/layouts/adaptive/use-window-size-classes).

I'm not convinced either by this, the three columns looks really cramped in phones, and could also in some tablets. I also prefer using the default libraries whenever possible, but for this case, it seems the default doesn't match what we need, WDYT about going the manual way here? Instead of using the SizeClass, we can simply wrap everything in a `BoxWithConstraints`, and use the `maxWidth` property to calculate the number of columns, and we can then have a column width that looks better.

JorgeMucientes committed:
Limit grid columns to the number of visible widgets

The number of columns in the grid layout should not exceed the number of available widgets.

JorgeMucientes committed:
Manually calculate number of columns based on  available width

Stop relying on Android material 3 WindowSizeClass categorization since the existing values doesn't work for us leading to a cramped UI with 3 columns on phones in landscape mode.

JorgeMucientes commented:
Thanks for the careful review @hichamboushaba. 
Those are really great suggestions for improving the experience. I've applied them both. Ready for another round 👍🏼

hichamboushaba reviewed (approved):
Great work 👏, this works really well, thanks @JorgeMucientes.

JorgeMucientes committed:
Merge branch 'trunk' into issue/11750-dashboard-big-screen-support
